File: wagerproof-ios-native/Wagerproof/Features/NFL/Components/NFLGameCard.swift
import SwiftUI
import WagerproofDesign
import WagerproofModels
import WagerproofServices

/// NFL game row rendered in the home Games feed list. Thin wrapper around
/// `GameRowCard` — populates the model ML edge from `homeAwayMlProb` vs
/// vegas implied probabilities. NFL's predictions table doesn't publish a
/// fair-total figure, so the O/U row surfaces direction + confidence only.
struct NFLGameCard: View {
    let game: NFLPrediction
    var onPress: () -> Void = {}
    @State private var slatePicks: [NFLSlatePickRow] = []

    var body: some View {
        GameRowCard(model: rowModel, onPress: onPress)
            .id("\(game.gameId)-mammoth-\(hasMammothPlay)")
            .task(id: game.gameId) {
                await loadSlatePicks()
            }
    }

    private var rowModel: GameRowCard.Model {
        // Canonical abbr + logo come from the `nfl_teams` reference table
        // (NFLTeamAssets, hydrated by the NFL fetch paths), with the static
        // identity map as the fallback.
        let awayAbbr = game.awayAb ?? NFLTeamAssets.abbr(for: game.awayTeam)
        let homeAbbr = game.homeAb ?? NFLTeamAssets.abbr(for: game.homeTeam)
        return GameRowCard.Model(
            id: game.id,
            league: "nfl",
            dateLabel: GameCardFormatting.formatCompactDate(game.kickoff ?? game.gameDate),
            timeLabel: GameCardFormatting.convertTimeToEST(game.kickoff ?? game.gameTime),
            away: GameRowCard.TeamSide(
                abbr: awayAbbr,
                initials: awayAbbr,
                moneyline: game.awayMl,
                spread: game.awaySpread,
                logoURL: NFLTeamAssets.logo(for: game.awayTeam),
                colors: NFLTeamColors.colorPair(for: game.awayTeam)
            ),
            home: GameRowCard.TeamSide(
                abbr: homeAbbr,
                initials: homeAbbr,
                moneyline: game.homeMl,
                spread: game.homeSpread,
                logoURL: NFLTeamAssets.logo(for: game.homeTeam),
                colors: NFLTeamColors.colorPair(for: game.homeTeam)
            ),
            overLine: game.overLine,
            mlEdge: nil,
            // Dry-run pipeline publishes a fair total (`fg_pred_total`); the
            // legacy pipeline publishes a direction probability instead.
            ouEdge: GameEdgeMath.ouEdge(
                modelFairTotal: game.predTotal,
                marketLine: game.overLine,
                ouResultProb: game.ouResultProb
            ),
            awayTeamFullName: game.awayTeam,
            homeTeamFullName: game.homeTeam,
            slatePicks: predictionPills,
            oddsBreakdown: oddsBreakdown,
            isMammoth: hasMammothPlay
        )
    }

    private var hasMammothPlay: Bool {
        if game.mammoth { return true }
        return slatePicks.contains { pick in
            pick.hasPlay == true
                && (pick.isMammoth == true || (pick.conviction ?? "").lowercased() == "mammoth")
        }
    }

    private var predictionPills: GameRowCard.SlatePicks {
        GameRowCard.SlatePicks(
            total: totalSlatePick,
            spread: spreadSlatePick,
            badges: slateBadges
        )
    }

    private var slateBadges: [GameRowCard.SlateBadge] {
        let highCount = slatePicks.filter { pick in
            pick.hasPlay == true && (pick.conviction ?? "").lowercased() == "high"
        }.count
        let signalCount = Set(slatePicks.flatMap(\.signalKeys)).count
        return GameRowCard.convictionBadges(
            hasMammoth: hasMammothPlay,
            highCount: highCount,
            signalCount: signalCount
        )
    }

    private var totalSlatePick: GameRowCard.SlateTotalPick? {
        guard let pick = slatePicks.first(where: { $0.cardGroup == "total" }),
              let direction = pickDirection(pick.pickSide ?? pick.pickLabel)
        else {
            guard let direction = pickDirection(game.fgTotalPick) else { return nil }
            return GameRowCard.SlateTotalPick(
                direction: direction,
                line: GameCardFormatting.roundToNearestHalf(game.fgTotalClose),
                color: direction == "UNDER" ? Color.appAccentRed : Color.appPrimary
            )
        }
        return GameRowCard.SlateTotalPick(
            direction: direction,
            line: GameCardFormatting.roundToNearestHalf(pick.bestLine ?? pick.vegasLine),
            color: direction == "UNDER" ? Color.appAccentRed : Color.appPrimary
        )
    }

    private var spreadSlatePick: GameRowCard.SlateSpreadPick? {
        guard let pick = slatePicks.first(where: { $0.cardGroup == "spread" }),
              let team = pick.pickTeam ?? teamName(forSide: game.fgSpreadPick)
        else {
            guard let team = teamName(forSide: game.fgSpreadPick) else { return nil }
            return GameRowCard.SlateSpreadPick(
                abbr: teamAbbr(team),
                logoURL: NFLTeamAssets.logo(for: team),
                line: spreadLineForTeam(team, line: game.fgSpreadClose),
                colors: NFLTeamColors.colorPair(for: team)
            )
        }
        return GameRowCard.SlateSpreadPick(
            abbr: teamAbbr(team),
            logoURL: NFLTeamAssets.logo(for: team),
            line: GameCardFormatting.formatSpread(pick.bestLine ?? pick.vegasLine),
            colors: NFLTeamColors.colorPair(for: team)
        )
    }

    /// Spread / Money / Total table — matches the MLB card layout. Over on the
    /// away row, Under on the home row.
    private var oddsBreakdown: GameRowCard.OddsBreakdown {
        let totalText = GameCardFormatting.roundToNearestHalf(game.overLine)
        let hasTotal = game.overLine != nil
        return GameRowCard.OddsBreakdown(
            away: GameRowCard.OddsBreakdown.Row(
                abbr: NFLTeamAssets.abbr(for: game.awayTeam),
                spread: GameCardFormatting.formatSpread(game.awaySpread),
                moneyline: GameCardFormatting.formatMoneyline(game.awayMl),
                total: hasTotal ? "O\(totalText)" : "—"
            ),
            home: GameRowCard.OddsBreakdown.Row(
                abbr: NFLTeamAssets.abbr(for: game.homeTeam),
                spread: GameCardFormatting.formatSpread(game.homeSpread),
                moneyline: GameCardFormatting.formatMoneyline(game.homeMl),
                total: hasTotal ? "U\(totalText)" : "—"
            )
        )
    }

    private func loadSlatePicks() async {
        guard (game.runId ?? "").localizedCaseInsensitiveContains("dryrun") else { return }
        let cfb = await CFBSupabase.shared.client
        guard let rows: [NFLSlatePickRow] = try? await cfb
            .from("nfl_dryrun_picks")
            .select("game_id,card_group,pick_team,pick_side,pick_label,best_line,vegas_line,conviction,is_mammoth,signal_keys,has_play,sort_order")
            .eq("game_id", value: game.gameId)
            .order("sort_order", ascending: true)
            .execute()
            .value
        else { return }
        slatePicks = rows
    }

    private func teamName(forSide side: String?) -> String? {
        let upper = (side ?? "").uppercased()
        if upper.contains("HOME") { return game.homeTeam }
        if upper.contains("AWAY") { return game.awayTeam }
        return nil
    }

    private func spreadLineForTeam(_ team: String, line: Double?) -> String {
        guard let line else { return "—" }
        return GameCardFormatting.formatSpread(team == game.homeTeam ? line : -line)
    }

    private func pickDirection(_ raw: String?) -> String? {
        let upper = (raw ?? "").uppercased()
        if upper.contains("UNDER") { return "UNDER" }
        if upper.contains("OVER") { return "OVER" }
        return nil
    }

    private func teamAbbr(_ team: String) -> String {
        if team == game.homeTeam, let homeAb = game.homeAb { return homeAb }
        if team == game.awayTeam, let awayAb = game.awayAb { return awayAb }
        return NFLTeamAssets.abbr(for: team)
    }

    private struct NFLSlatePickRow: Decodable, Sendable {
        let cardGroup: String?
        let pickTeam: String?
        let pickSide: String?
        let pickLabel: String?
        let bestLine: Double?
        let vegasLine: Double?
        let conviction: String?
        let isMammoth: Bool?
        let signalKeys: [String]
        let hasPlay: Bool?

        enum CodingKeys: String, CodingKey {
            case cardGroup = "card_group"
            case pickTeam = "pick_team"
            case pickSide = "pick_side"
            case pickLabel = "pick_label"
            case bestLine = "best_line"
            case vegasLine = "vegas_line"
            case conviction
            case isMammoth = "is_mammoth"
            case signalKeys = "signal_keys"
            case hasPlay = "has_play"
        }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            cardGroup = try c.decodeIfPresent(String.self, forKey: .cardGroup)
            pickTeam = try c.decodeIfPresent(String.self, forKey: .pickTeam)
            pickSide = try c.decodeIfPresent(String.self, forKey: .pickSide)
            pickLabel = try c.decodeIfPresent(String.self, forKey: .pickLabel)
            bestLine = try c.decodeIfPresent(Double.self, forKey: .bestLine)
            vegasLine = try c.decodeIfPresent(Double.self, forKey: .vegasLine)
            conviction = try c.decodeIfPresent(String.self, forKey: .conviction)
            isMammoth = try c.decodeIfPresent(Bool.self, forKey: .isMammoth)
            signalKeys = (try? c.decodeIfPresent(FlexibleStringList.self, forKey: .signalKeys))?.values ?? []
            hasPlay = try c.decodeIfPresent(Bool.self, forKey: .hasPlay)
        }
    }

    private struct FlexibleStringList: Decodable, Sendable {
        let values: [String]

        init(from decoder: Decoder) throws {
            let c = try decoder.singleValueContainer()
            if let array = try? c.decode([String].self) {
                values = array.filter { !$0.isEmpty }
            } else if let string = try? c.decode(String.self) {
                if let data = string.data(using: .utf8),
                   let parsed = try? JSONDecoder().decode([String].self, from: data) {
                    values = parsed.filter { !$0.isEmpty }
                } else {
                    values = string
                        .split(separator: ",")
                        .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                        .filter { !$0.isEmpty }
                }
            } else {
                values = []
            }
        }
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/nfl/NFLGameCard.kt
package com.wagerproof.app.features.nfl

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.wagerproof.app.features.gamecards.GameCardFormatting
import com.wagerproof.app.features.gamecards.GameEdgeMath
import com.wagerproof.app.features.gamecards.GameRowCard
import com.wagerproof.app.features.gamecards.GameRowCardModel
import com.wagerproof.app.features.gamecards.TeamColorPair
import com.wagerproof.app.features.props.nflTeamColors
import com.wagerproof.core.models.NFLPrediction
import com.wagerproof.core.models.NFLTeamAssets
import java.util.Locale
import kotlin.math.floor

/**
 * NFL game row for the home Games feed — a thin adapter over the shared
 * [GameRowCard], mirroring iOS `NFLGameCard`. Dry-run cards hydrate the
 * authoritative `nfl_dryrun_picks` rows; merged game fields remain the loading,
 * legacy, and missing-row fallback.
 */
@Composable
fun NFLGameCard(
    game: NFLPrediction,
    onPress: () -> Unit,
    modifier: Modifier = Modifier,
) {
    var picks by remember(game.gameId) { mutableStateOf<List<NFLDryrunPickRow>>(emptyList()) }
    LaunchedEffect(game.gameId, game.runId) {
        if ((game.runId ?: "").contains("dryrun", ignoreCase = true)) {
            loadNFLSlatePicksResult(game.gameId).onSuccess { picks = it }
        }
    }

    val awayAbbr = game.awayAb ?: NFLTeamAssets.abbr(game.awayTeam)
    val homeAbbr = game.homeAb ?: NFLTeamAssets.abbr(game.homeTeam)

    val model = GameRowCardModel(
        id = game.id,
        league = "nfl",
        dateLabel = GameCardFormatting.formatCompactDate(game.kickoff ?: game.gameDate),
        timeLabel = GameCardFormatting.convertTimeToEST(game.kickoff ?: game.gameTime),
        away = GameRowCardModel.TeamSide(
            abbr = awayAbbr,
            initials = awayAbbr,
            moneyline = game.awayMl,
            spread = game.awaySpread,
            logoURL = NFLTeamAssets.logo(game.awayTeam),
            colors = nflColorPair(game.awayTeam),
        ),
        home = GameRowCardModel.TeamSide(
            abbr = homeAbbr,
            initials = homeAbbr,
            moneyline = game.homeMl,
            spread = game.homeSpread,
            logoURL = NFLTeamAssets.logo(game.homeTeam),
            colors = nflColorPair(game.homeTeam),
        ),
        overLine = game.overLine,
        mlEdge = null,
        // Dry-run pipeline publishes a fair total (pred_total); the legacy
        // pipeline publishes a direction probability instead.
        ouEdge = GameEdgeMath.ouEdge(
            modelFairTotal = game.predTotal,
            marketLine = game.overLine,
            ouResultProb = game.ouResultProb,
        ),
        awayTeamFullName = game.awayTeam,
        homeTeamFullName = game.homeTeam,
        slatePicks = nflSlatePicks(game, picks),
        oddsBreakdown = oddsBreakdown(game),
        isMammoth = nflHasMammothPlay(game, picks),
    )

    GameRowCard(model = model, onPress = onPress, modifier = modifier)
}

/** [TeamColorPair] over the authoritative 32-team brand map. */
private fun nflColorPair(team: String): TeamColorPair {
    val (primary, secondary) = nflTeamColors(team)
    return TeamColorPair(primary, secondary)
}

internal fun nflHasMammothPlay(game: NFLPrediction, picks: List<NFLDryrunPickRow>): Boolean =
    game.mammoth || picks.any { pick ->
        pick.hasPlay == true && (pick.isMammoth == true || pick.conviction.equals("mammoth", ignoreCase = true))
    }

internal fun nflSlatePicks(game: NFLPrediction, picks: List<NFLDryrunPickRow>): GameRowCardModel.SlatePicks {
    val totalPick = picks.firstOrNull { it.cardGroup == "total" }
    val totalDir = pickDirection(totalPick?.pickSide ?: totalPick?.pickLabel ?: game.fgTotalPick)
    val totalLine = totalPick?.let { it.bestLine ?: it.vegasLine } ?: game.fgTotalClose
    val totalLabel = totalDir?.let { "O/U $it ${fmtHalf(totalLine)}" }

    val spreadPick = picks.firstOrNull { it.cardGroup == "spread" }
    val spreadTeam = spreadPick?.pickTeam
        ?: teamName(game, spreadPick?.pickSide)
        ?: teamName(game, game.fgSpreadPick)
    val spreadLogo = spreadTeam?.let { NFLTeamAssets.logo(it) }
    val spreadLabel = spreadTeam?.let {
        val line = if (spreadPick != null) {
            spreadPick.bestLine ?: spreadPick.vegasLine
        } else {
            val isHome = it == game.homeTeam
            game.fgSpreadClose?.let { v -> if (isHome) v else -v }
        }
        GameCardFormatting.formatSpread(line)
    }

    val highCount = if (picks.isNotEmpty()) {
        picks.count { it.hasPlay == true && it.conviction.equals("high", ignoreCase = true) }
    } else {
        game.convictionSummary?.plays.orEmpty().count { (it.conviction ?: "").lowercase(Locale.US) == "high" }
    }
    val signalCount = if (picks.isNotEmpty()) {
        picks.flatMap { it.signalKeys }.filter(String::isNotBlank).toSet().size
    } else {
        game.flagsActive ?: 0
    }

    return GameRowCardModel.SlatePicks(
        totalIsOver = totalDir?.let { it == "OVER" },
        totalLabel = totalLabel,
        spreadLogoURL = spreadLogo,
        spreadLabel = spreadLabel,
        hasMammoth = nflHasMammothPlay(game, picks),
        highCount = highCount,
        signalCount = signalCount,
    )
}

private fun oddsBreakdown(game: NFLPrediction): GameRowCardModel.OddsBreakdown {
    val hasTotal = game.overLine != null
    val totalText = fmtHalf(game.overLine)
    return GameRowCardModel.OddsBreakdown(
        awaySpread = GameCardFormatting.formatSpread(game.awaySpread),
        homeSpread = GameCardFormatting.formatSpread(game.homeSpread),
        awayML = GameCardFormatting.formatMoneyline(game.awayMl),
        homeML = GameCardFormatting.formatMoneyline(game.homeMl),
        awayTotal = if (hasTotal) "O$totalText" else "—",
        homeTotal = if (hasTotal) "U$totalText" else "—",
    )
}

private fun teamName(game: NFLPrediction, side: String?): String? {
    val upper = (side ?: "").uppercase(Locale.US)
    return when {
        upper.contains("HOME") -> game.homeTeam
        upper.contains("AWAY") -> game.awayTeam
        else -> null
    }
}

private fun fmtHalf(value: Double?): String {
    val rounded = GameCardFormatting.roundToNearestHalf(value) ?: return "—"
    return if (floor(rounded) == rounded) rounded.toInt().toString() else String.format(Locale.US, "%.1f", rounded)
}

private fun pickDirection(raw: String?): String? {
    val u = (raw ?: "").uppercase(Locale.US)
    return when {
        u.contains("UNDER") -> "UNDER"
        u.contains("OVER") -> "OVER"
        else -> null
    }
}
